components/brave_wallet/browser/brave_wallet_pin_service.cc

```diff
@@ -0,0 +1,778 @@
+      ResolvePinItems(service, token);
+
+  if (!cids) {
+    SetTokenStatus(service, std::move(token),
+                   mojom::TokenPinStatusCode::STATUS_PINNING_IN_PROGRESS,
+                   nullptr);
+    std::move(callback).Run(true, nullptr);
+    return;
+  }
+
+  auto path = GetPath(absl::nullopt, token);
+  if (!path) {
+    std::move(callback).Run(
+        false,
+        mojom::PinError::New(mojom::WalletPinServiceErrorCode::ERR_WRONG_TOKEN,
+                             "Wrong token data"));
+    return;
+  }
+
+  if (!service) {
+    local_pin_service_->ValidatePins(
+        path.value(), cids.value(),
+        base::BindOnce(&BraveWalletPinService::OnTokenValidated,
+                       base::Unretained(this), service, std::move(callback),
+                       std::move(token)));
+  } else {
+    // Remote pinning not implemented yet
+    std::move(callback).Run(false, nullptr);
+  }
+}
+
+void BraveWalletPinService::IsLocalNodeRunning(
+    IsLocalNodeRunningCallback callback) {
+  std::move(callback).Run(ipfs_service_->IsDaemonLaunched());
+}
+
+void BraveWalletPinService::OnIpfsLaunched(bool result, int64_t pid) {
+  for (const auto& observer : observers_) {
+    observer->OnLocalNodeStatusChanged(result);
+  }
+}
+
+void BraveWalletPinService::OnIpfsShutdown() {
+  for (const auto& observer : observers_) {
+    observer->OnLocalNodeStatusChanged(false);
+  }
+}
+
+void BraveWalletPinService::MarkAsPendingForPinning(
+    const mojom::BlockchainTokenPtr& token,
+    const absl::optional<std::string>& service) {
+  SetTokenStatus(service, token,
+                 mojom::TokenPinStatusCode::STATUS_PINNING_PENDING, nullptr);
+}
+
+void BraveWalletPinService::MarkAsPendingForUnpinning(
+    const mojom::BlockchainTokenPtr& token,
+    const absl::optional<std::string>& service) {
+  SetTokenStatus(service, token,
+                 mojom::TokenPinStatusCode::STATUS_UNPINNING_PENDING, nullptr);
+}
+
+void BraveWalletPinService::AddPin(BlockchainTokenPtr token,
+                                   const absl::optional<std::string>& service,
+                                   AddPinCallback callback) {
+  if (!token->is_nft) {
+    auto pin_error = mojom::PinError::New(
+        mojom::WalletPinServiceErrorCode::ERR_WRONG_TOKEN, "Token is not nft");
+
+    VLOG(1) << "Token is not nft";
+    FinishAddingWithResult(service, token, false, std::move(pin_error),
+                           std::move(callback));
+    return;
+  }
+
+  auto token_status = GetTokenStatus(service, token);
+  if (token_status &&
+      token_status->code == mojom::TokenPinStatusCode::STATUS_PINNED) {
+    auto pin_error = mojom::PinError::New(
+        mojom::WalletPinServiceErrorCode::ERR_ALREADY_PINNED, "Already pinned");
+
+    FinishAddingWithResult(service, token, true, std::move(pin_error),
+                           std::move(callback));
+    return;
+  }
+
+  json_rpc_service_->GetERC721Metadata(
```

Comment:
Does our MVP only include ERC721? Please open follow-ups for others (ERC1155, Solana NFT).

Reply:
And we should only call this when is_erc721 is true, and return error result such as NFT standard unsupported at the moment when is_erc721 is false.